### PR TITLE
Add debug header for diagnostic tools

### PR DESCRIPTION
**Pull request by davmason**
This change adds the concept of a "Debug Header" which allows diagnostic tooling to get information about the runtime types and layout without being tightly coupled to the runtime like the DAC is for coreclr. It is a lightweight self describing metadata format focused on types and their members, globals, and build defines.

### CODE REVIEW

File: src/coreclr/nativeaot/Runtime/DebugHeader.cpp [hunk 1]
```diff
@@ -0,0 +1,275 @@
...
+
+    DefineEntry *DefinesList = nullptr;
+};
+
+extern "C" NativeAOTRuntimeDebugHeader g_NativeAOTRuntimeDebugHeader = {};
+
+#define MAKE_DEBUG_ENTRY(TypeName, FieldName, Value)                                                                        \
+    do                                                                                                                      \
+    {                                                                                                                       \
+        currentType = new (nothrow) DebugTypeEntry{ previousType, #TypeName, #FieldName, Value };                             \
```
Comment (MichalStrehovsky) [inline src/coreclr/nativeaot/Runtime/DebugHeader.cpp:115]: I'm wondering if we could pre-initialize most of these in the data region. The extra `new`s are probably not measurably expensive, but if we can avoid doing doing extra work at startup, it's better to avoid it.

File: src/coreclr/nativeaot/Runtime/DebugHeader.cpp [hunk 2]
```diff
@@ -0,0 +1,275 @@
...
+    MAKE_DEBUG_FIELD_ENTRY(ThreadBuffer, m_pThreadStressLog);
+
+    // EEThreadID is forward declared and not available
+    MAKE_DEBUG_ENTRY(EEThreadID, SIZEOF, sizeof(void*));
+    MAKE_DEBUG_ENTRY(EEThreadID, m_FiberPtrId, 0);
+
+    MAKE_SIZE_ENTRY(EEType);
+    MAKE_DEBUG_FIELD_ENTRY(EEType, m_uBaseSize);
+    MAKE_DEBUG_FIELD_ENTRY(EEType, m_usComponentSize);
+    MAKE_DEBUG_FIELD_ENTRY(EEType, m_usFlags);    MAKE_DEBUG_ENTRY(EEType, m_pBaseType, offsetof(EEType, m_RelatedType) + offsetof(EEType::RelatedTypeUnion, m_pBaseType));
```
Comment (MichalStrehovsky) [inline src/coreclr/nativeaot/Runtime/DebugHeader.cpp:198]: It would be good to capture somewhere which flags the debugger is going to depend on. We have been changing these in the past.

File: src/coreclr/nativeaot/Runtime/DebugHeader.cpp (hunk 1 above)
Comment (jkotas) [reply src/coreclr/nativeaot/Runtime/DebugHeader.cpp:115]: +1 I was about to write the same feedback.

File: src/coreclr/nativeaot/Runtime/DebugHeader.cpp (hunk 2 above)
Comment (davmason) [reply src/coreclr/nativeaot/Runtime/DebugHeader.cpp:198]: The flags are used for !verifyheap, the logic is copied from [EEType::Validate](https://github.com/dotnet/runtimelab/blob/99dbab06575ab15fac0fbfd53df1ecd863914fbe/src/coreclr/nativeaot/Runtime/eetype.cpp#L19) for object validation. 

I assumed that the flags were stable, I can add some static asserts though that check that they change and instruct to bump the debug header version if they are different.

File: src/coreclr/nativeaot/Runtime/DebugHeader.cpp (hunk 1 above)
Comment (davmason) [reply src/coreclr/nativeaot/Runtime/DebugHeader.cpp:115]: I originally intended to have it be all at compile time. The data is just offsets and build defines so it is available at  compile time.  I spent some time a while back trying to get it to work with constexpr functions, but the main issue I kept running in to was how to get a dynamically sized array that works with constexpr. I suppose I could have a statically sized array and just have to keep it in sync with the number of entries.

I can take a look at it again and see if I can come up with something. If there's any prior art I could follow please let me know.

Comment (davmason) [reply src/coreclr/nativeaot/Runtime/DebugHeader.cpp]: I wasn't able to come up with a way to embed them at compile time like I would like, but I at least got rid of the allocations. Now there is an array and at runtime the values are assigned to the array without allocation.

File: src/coreclr/nativeaot/Runtime/DebugHeader.cpp (hunk 2 above)
Comment (davmason) [reply src/coreclr/nativeaot/Runtime/DebugHeader.cpp]: Asserts are added now

**Comment by jkotas** [comment]
Could you please take a look at the build breaks?

**Comment by davmason** [comment]
Build breaks are fixed and I tested the SOS changed on Linux as well.

File: src/coreclr/nativeaot/Runtime/DebugHeader.cpp [hunk 3]
```diff
@@ -0,0 +1,294 @@
...
+    
+    // Reserved - Currently it only serves as alignment padding for the pointers which
+    // follow but future usage will be considered a back-compatible change.
+    const uint32_t ReservedPadding3 = 0;
+
+    // An array that contains pointers to important globals
+    GlobalValueEntry GlobalEntries[GLOBALS_ARRAY_SIZE];
+};
+
+extern "C" NativeAOTRuntimeDebugHeader g_NativeAOTRuntimeDebugHeader = {};
```
Comment (jkotas) [inline src/coreclr/nativeaot/Runtime/DebugHeader.cpp:118]: Would it make sense to have non-NativeAOT specific name for this? E.g. `DotNetRuntimeDebugHeader`.

I would love to see this unified accross all runtime variants eventually.

File: src/coreclr/nativeaot/Runtime/DebugHeader.cpp [hunk 4]
```diff
@@ -0,0 +1,294 @@
...
+    MAKE_DEBUG_FIELD_ENTRY(StressMsg, fmtOffsCArgs);
+    MAKE_DEBUG_FIELD_ENTRY(StressMsg, facility);
+    MAKE_DEBUG_FIELD_ENTRY(StressMsg, timeStamp);
+    MAKE_DEBUG_FIELD_ENTRY(StressMsg, args);
+
+    MAKE_SIZE_ENTRY(Object);
+    MAKE_DEBUG_FIELD_ENTRY(Object, m_pEEType);
+
+    MAKE_SIZE_ENTRY(Array);
+    MAKE_DEBUG_FIELD_ENTRY(Array, m_Length);
```
Comment (jkotas) [inline src/coreclr/nativeaot/Runtime/DebugHeader.cpp:264]: I do not think we need to have the System.Object and array layout configurable. They are pretty much set in stone.

File: src/coreclr/nativeaot/Runtime/DebugHeader.cpp [hunk 5]
```diff
@@ -0,0 +1,294 @@
...
+
+    MAKE_SIZE_ENTRY(ThreadBuffer);
+    MAKE_DEBUG_FIELD_ENTRY(ThreadBuffer, m_pNext);
+    MAKE_DEBUG_FIELD_ENTRY(ThreadBuffer, m_rgbAllocContextBuffer);
+    MAKE_DEBUG_FIELD_ENTRY(ThreadBuffer, m_threadId);
+    MAKE_DEBUG_FIELD_ENTRY(ThreadBuffer, m_pThreadStressLog);
+
+    // EEThreadID is forward declared and not available
+    MAKE_DEBUG_ENTRY(EEThreadID, SIZEOF, sizeof(void*));
+    MAKE_DEBUG_ENTRY(EEThreadID, m_FiberPtrId, 0);
```
Comment (jkotas) [inline src/coreclr/nativeaot/Runtime/DebugHeader.cpp:206]: Is this right for non-Windows?

**Review by jkotas — APPROVED** [review]
LGTM otherwise. Thanks!

File: src/coreclr/nativeaot/Runtime/DebugHeader.cpp [hunk 6]
```diff
@@ -0,0 +1,294 @@
...
+    // The length of the DebugTypeEntries array
+    #define DEBUG_TYPES_ARRAY_SIZE 101
+    const uint32_t DebugTypeEntriesArraySize = DEBUG_TYPES_ARRAY_SIZE;
+
+    // Reserved - Currently it only serves as alignment padding for the pointers which
+    // follow but future usage will be considered a back-compatible change.
+    const uint32_t ReservedPadding2 = 0;
+
+    // An array describing important types and their offsets
+    DebugTypeEntry DebugTypeEntries[DEBUG_TYPES_ARRAY_SIZE];
```
Comment (MichalStrehovsky) [inline src/coreclr/nativeaot/Runtime/DebugHeader.cpp:104]: I'm wondering whether it would be useful to have a little bit more extensibility in this structure. The previous design had some pointer-typed fields that allowed to do things with it.

I was thinking that if we're ever in a situation where we need to capture information in this structure that is not known at the time of compiling the native runtime bits, but becomes known when we do the AOT compilation of the managed code, we could have the AOT compiler generate the necessary part of the data structure. But with DebugHeader in this shape, there's no way to append those records to this structure (we would have to reformat it at startup).

If instead of storing these entries inline in the header, we put a pointer here to the inline array (still stored statically), and if instead of storing the count here, we would have a convention that the array ends with a single all-null entry, this would allow enough extensibility that we can append to this list with the platform linker (i.e. the compiler can generate additional entries and the linker would lay it out in the appropriate spot so that it shows up continuously in the image).

File: src/coreclr/nativeaot/Runtime/DebugHeader.cpp [hunk 7]
```diff
@@ -0,0 +1,294 @@
...
+    static_assert(EEType::Flags::HasFinalizerFlag       == 0x0010, "SOS has a hard coded dependency on the values of EEType::Flags. If you change these values you must bump major_version_number.");
+    static_assert(EEType::Flags::HasPointersFlag        == 0x0020, "SOS has a hard coded dependency on the values of EEType::Flags. If you change these values you must bump major_version_number.");
+    static_assert(EEType::Flags::GenericVarianceFlag    == 0x0080, "SOS has a hard coded dependency on the values of EEType::Flags. If you change these values you must bump major_version_number.");
+    static_assert(EEType::Flags::OptionalFieldsFlag     == 0x0100, "SOS has a hard coded dependency on the values of EEType::Flags. If you change these values you must bump major_version_number.");
+    static_assert(EEType::Flags::IsGenericFlag          == 0x0400, "SOS has a hard coded dependency on the values of EEType::Flags. If you change these values you must bump major_version_number.");
+    static_assert(EEType::Flags::ElementTypeMask        == 0xf800, "SOS has a hard coded dependency on the values of EEType::Flags. If you change these values you must bump major_version_number.");
```
Comment (MichalStrehovsky) [inline src/coreclr/nativeaot/Runtime/DebugHeader.cpp:292]: Do we really have a dependency on `IsDynamicType` and `OptionalFieldsFlag`?

File: src/coreclr/nativeaot/Bootstrap/main.cpp [hunk 8]
```diff
@@ -164,6 +166,9 @@ static int InitializeRuntime()
 
     InitializeModules(osModule, __modules_a, (int)((__modules_z - __modules_a)), (void **)&c_classlibFunctions, _countof(c_classlibFunctions));
 
+    // Populate the values needed for debugging
```
Comment (MichalStrehovsky) [inline src/coreclr/nativeaot/Bootstrap/main.cpp:169]: Wondering if we could just call this from `RhInitialize`.

**Review by MichalStrehovsky — APPROVED** [review]
Thank you!

File: src/coreclr/nativeaot/Runtime/DebugHeader.cpp (hunk 6 above)
Comment (jkotas) [reply src/coreclr/nativeaot/Runtime/DebugHeader.cpp:104]: `g_gcDacGlobals` in the current list demonstrates similar problem as the AOT compiler generated records.
 
I think that this flat list may over time evolve into a list of individually versioned contracts that are defined locally by each component (StressLog, Threading, GC, AOT compiler/typesystem, ...).

File: src/coreclr/nativeaot/Runtime/DebugHeader.cpp (hunk 3 above)
Comment (davmason) [reply src/coreclr/nativeaot/Runtime/DebugHeader.cpp:118]: I also would like to see this approach unified across all the runtimes, I was originally thinking that they would have their own contracts but there's no reason not to have the same name for the type of the contract itself.

I would like to keep the export name as "g_NativeAOTRuntimeDebugHeader" though so we can differentiate runtimes by just checking the export name. So then coreclr can have g_CoreCLRRuntimeDebugHeader, etc.

File: src/coreclr/nativeaot/Runtime/DebugHeader.cpp (hunk 5 above)
Comment (davmason) [reply src/coreclr/nativeaot/Runtime/DebugHeader.cpp:206]: I thought it was. In coreclr\vm\util.hpp EEThreadId just has the one member of m_FiberPtrId.

It doesn't really matter for this work anyways, none of the commands need to access EEThreadID internals. I just need to read from ThreadBuffer and can delete these entries

File: src/coreclr/nativeaot/Runtime/DebugHeader.cpp (hunk 7 above)
Comment (davmason) [reply src/coreclr/nativeaot/Runtime/DebugHeader.cpp:292]: No, I was just on auto pilot here

File: src/coreclr/nativeaot/Bootstrap/main.cpp (hunk 8 above)
Comment (davmason) [reply src/coreclr/nativeaot/Bootstrap/main.cpp:169]: Sure!

File: src/coreclr/nativeaot/Runtime/DebugHeader.cpp (hunk 3 above)
Comment (jkotas) [reply src/coreclr/nativeaot/Runtime/DebugHeader.cpp]: We like to talk about NativeAOT project as a flavor of CoreCLR runtime. The place we want to get to is there is a continuum of runtime configurations to choose from, not distinct runtimes.

Comment (davmason) [reply src/coreclr/nativeaot/Runtime/DebugHeader.cpp]: Is the concern over API surface area? For a tool like SOS which is inspecting runtime level data structures we will have to be aware which runtime flavor we are running against for things like whether to look up the EEType in the symbols file, or to use CLR metadata to get type information.

I'm happy to change it to g_DotNetRuntimeDebugHeader, but then I would have to either differentiate on the Cookie or some define we put in the contract. That doesn't feel meaningfully different to me other than the API surface.

Comment (jkotas) [reply src/coreclr/nativeaot/Runtime/DebugHeader.cpp]: The places that use this should differentiate on contract type, not based on runtime type.

For example, stresslog or GC contract is same between all runtimes, no need to differentiate today.

For type system contract, there are multiple choices - e.g. try JIT-based type system contract first, and if it is not available try no-JIT AOT-based contract second.

File: src/coreclr/nativeaot/Runtime/ObjectLayout.h [hunk 9]
```diff
@@ -37,9 +37,13 @@ typedef DPTR(class EEType) PTR_EEType;
 class MethodTable;
 
 //-------------------------------------------------------------------------------------------------
+
+extern "C" void PopulateDebugHeaders();
```
Comment (MichalStrehovsky) [inline src/coreclr/nativeaot/Runtime/ObjectLayout.h:41]: It doesn't look like the changes to this file are necessary anymore.

File: src/coreclr/nativeaot/Runtime/threadstore.cpp [hunk 10]
```diff
@@ -542,4 +542,5 @@ Boolean ThreadStore::GetExceptionsForCurrentThread(Array* pOutputArray, int32_t*
     *pWrittenCountOut = countWritten;
     return Boolean_false;
 }
+
```
Comment (MichalStrehovsky) [inline src/coreclr/nativeaot/Runtime/threadstore.cpp:545]: Nit: unnecessary whitespace change

**Review by MichalStrehovsky — APPROVED** [review]
Thank you!

File: src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.targets [hunk 11]
```diff
@@ -281,6 +281,7 @@ The .NET Foundation licenses this file to you under the MIT license.
       <CustomLinkerArg Include="-o &quot;$(NativeBinary)&quot;" Condition="'$(TargetOS)' != 'windows'" />
       <CustomLinkerArg Include="/OUT:&quot;$(NativeBinary)&quot;" Condition="'$(TargetOS)' == 'windows'" />
       <CustomLinkerArg Include="/DEF:&quot;$(ExportsFile)&quot;" Condition="'$(TargetOS)' == 'windows' and $(ExportsFile) != ''" />
+      <CustomLinkerArg Include="/EXPORT:g_DotNetRuntimeDebugHeader,DATA" Condition="'$(OS)' == 'Windows_NT'" />
```
Comment (MichalStrehovsky) [inline src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.targets:284]: Btw, would the debugger be able to find this symbol in the PDB? Do we need to export it?

If it's exported, a lot of people are going to see it so it might be "nicer" to go with Jan's original suggestion of naming it `DotNetRuntimeDebugHeader` without the `g_`. If we're expecting to keep this as our debugging future, we might as well make it not look like an accidentally exported implementation detail. It's more of a nit of course.

Comment (davmason) [reply src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.targets:284]: If it wasn't exported I think the debugger could find it, but there are other tools outside of a debugger that we would want to support, such as the dotnet-dump tool. It's a lot easier to find as an export than having to be able to load a symbol file. I removed the "g_" from the name.

File: src/coreclr/nativeaot/Runtime/threadstore.cpp (hunk 10 above)
Comment (davmason) [reply src/coreclr/nativeaot/Runtime/threadstore.cpp:545]: Removed!

File: src/coreclr/nativeaot/Runtime/ObjectLayout.h (hunk 9 above)
Comment (davmason) [reply src/coreclr/nativeaot/Runtime/ObjectLayout.h:41]: Removed those too

File: src/coreclr/nativeaot/Runtime/DebugHeader.cpp [hunk 12]
```diff
@@ -0,0 +1,279 @@
...
+struct DotNetRuntimeDebugHeader
+{
+    // The cookie serves as a sanity check against process corruption or being requested
+    // to treat some other non-.Net module as though it did contain the coreRT runtime.
+    // It can also be changed if we want to make a breaking change so drastic that
+    // earlier debuggers should treat the module as if it had no .Net runtime at all.
+    // If the cookie is valid a debugger is safe to assume the Major/Minor version fields
+    // will follow, but any contents beyond that depends on the version values.
+    // The cookie value is currently set to 0x4E 0x41 0x44 0x48 (NADH in ascii)
+    const uint8_t Cookie[4] = { 0x4E, 0x41, 0x44, 0x48 };
```
Comment (noahfalk) [inline src/coreclr/nativeaot/Runtime/DebugHeader.cpp:77]: Nit: Does NADH stand for 'Native AOT Debug Header'? If we changed the name elsewhere maybe this should change too for naming consistency.

File: src/coreclr/nativeaot/Runtime/DebugHeader.cpp [hunk 13]
```diff
@@ -0,0 +1,279 @@
...
+
+    void *g_stressLog = &StressLog::theLog;
+    MAKE_GLOBAL_ENTRY(g_stressLog);
+
+    // Some DAC functions need to know the module base address, easiest way is with
+    // the HANDLE to our module which is the base address.
+    HANDLE moduleBaseAddress = PalGetModuleHandleFromPointer((void *)&PopulateDebugHeaders);
+    MAKE_GLOBAL_ENTRY(moduleBaseAddress);
+
+    static_assert(EEType::Flags::EETypeKindMask         == 0x0003, "SOS has a hard coded dependency on the values of EEType::Flags. If you change these values you must bump major_version_number.");
```
Comment (noahfalk) [inline src/coreclr/nativeaot/Runtime/DebugHeader.cpp:271]: Nit: I would replace 'SOS' with a more generic 'Debugging data contracts'. SOS may be the first user of this but it probably won't be the last.

File: src/coreclr/nativeaot/Runtime/DebugHeader.cpp [hunk 14]
```diff
@@ -0,0 +1,279 @@
...
+// This size should be one bigger than the number of entries since a null entry
+// signifies the end of the array.
+static constexpr size_t GlobalEntriesArraySize = 6;
+static GlobalValueEntry s_GlobalEntries[GlobalEntriesArraySize];
+
+// This structure is part of a in-memory serialization format that is used by diagnostic tools to
+// reason about the runtime. As a contract with our diagnostic tools it must be kept up-to-date
+// by changing the MajorVersion when breaking changes occur. If you are changing the runtime then
+// you are responsible for understanding what changes are breaking changes. You can do this by
+// reading the specification (Documentation\design-docs\diagnostics\ProcessMemoryFormatSpec.md)
```
Comment (noahfalk) [inline src/coreclr/nativeaot/Runtime/DebugHeader.cpp:50]: Lacking the doc we should probably remove this part of the comment : )

File: src/coreclr/nativeaot/Runtime/DebugHeader.cpp [hunk 15]
```diff
@@ -0,0 +1,279 @@
...
+extern "C" void PopulateDebugHeaders()
+{
+    size_t currentDebugPos = 0;
+    size_t currentGlobalPos = 0;
+
+    ZeroMemory(s_DebugEntries, DebugTypeEntriesArraySize);
+    ZeroMemory(s_GlobalEntries, GlobalEntriesArraySize);
+
+    DotNetRuntimeDebugHeader.DebugTypeEntries = &s_DebugEntries;
+    DotNetRuntimeDebugHeader.GlobalEntries = &s_GlobalEntries;
```
Comment (noahfalk) [inline src/coreclr/nativeaot/Runtime/DebugHeader.cpp:145]: We could get rare race conditions if we publish the pointer to the table before it is initialized. Defining these pointers as volatile and setting them after the table is initialized should eliminate that possibility.

**Review by noahfalk — APPROVED** [review]
A few small things but overall LGTM!

File: src/coreclr/nativeaot/Runtime/DebugHeader.cpp (hunk 12 above)
Comment (davmason) [reply src/coreclr/nativeaot/Runtime/DebugHeader.cpp:77]: That is definitely what it stands for 😄. I can change it to DNDH for DotNetDebugHeader.

File: src/coreclr/nativeaot/Runtime/DebugHeader.cpp (hunk 15 above)
Comment (davmason) [reply src/coreclr/nativeaot/Runtime/DebugHeader.cpp:145]: Sure, I can make that change

**Comment by davmason** [comment]
Thanks everybody for reviewing. At this point I believe all feedback is accounted for and I am planning on merging later tonight.
